File: 2_queue/单调队列Monoqueue/MaxDequeue.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable no-console */
/* eslint-disable no-param-reassign */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
// 用两个栈模拟一个双端队列，注意到每个栈的前缀最小值是可以 O(1) 修改和询问的
// 当某个栈空的时候将另一个栈分一半给这个已经空的栈（即暴力重构）

interface IDeque<E = unknown> {
  append(item: E): void
  appendLeft(item: E): void
  pop(): E | undefined
  popLeft(): E | undefined
  at(index: number): E | undefined
  readonly size: number
}

interface IStack<E = unknown> {
  append(item: E): void
  pop(): E | undefined
  at(index: number): E | undefined
  readonly size: number
}

type CompareFunc<E> = (a: E, b: E) => number

class MaxDeque<E extends { value: V }, V = E['value']> implements IDeque<E> {
  private readonly _left: _MaxStack<E, V>
  private readonly _right: _MaxStack<E, V>
  private readonly _compareValue: CompareFunc<V>

  constructor(compareValue: CompareFunc<V> = (a: any, b: any) => a - b) {
    this._left = new _MaxStack(compareValue)
    this._right = new _MaxStack(compareValue)
    this._compareValue = compareValue
  }

  append(item: E): void {
    this._right.append(item)
  }

  appendLeft(item: E): void {
    this._left.append(item)
  }

  pop(): E | undefined {
    if (this._right.size > 0) {
      return this._right.pop()
    }
    const tmp: E[] = []
    const n = this._left.size
    for (let i = 0; i < n; i++) {
      tmp.push(this._left.pop()!)
    }
    const half = n >> 1
    for (let i = half - 1; i >= 0; i--) {
      this._left.append(tmp[i])
    }
    for (let i = half; i < n; i++) {
      this._right.append(tmp[i])
    }
    return this._right.pop()
  }

  popLeft(): E | undefined {
    if (this._left.size > 0) {
      return this._left.pop()
    }
    const tmp: E[] = []
    const n = this._right.size
    for (let i = 0; i < n; i++) {
      tmp.push(this._right.pop()!)
    }
    const half = n >> 1
    for (let i = half - 1; i >= 0; i--) {
      this._right.append(tmp[i])
    }
    for (let i = half; i < n; i++) {
      this._left.append(tmp[i])
    }
    return this._left.pop()
  }

  at(index: number): E | undefined {
    const n = this.size
    if (index < 0) {
      index += n
    }
    if (index < 0 || index >= n) {
      return undefined
    }
    const leftSize = this._left.size
    if (index < leftSize) {
      return this._left.at(leftSize - index - 1)
    }
    return this._right.at(index - leftSize)
  }

  toString(): string {
    const sb: string[] = []
    const n = this.size
    for (let i = 0; i < n; i++) {
      sb.push(JSON.stringify(this.at(i)))
    }
    return `MaxDeque(${sb.join(', ')})`
  }

  get size(): number {
    return this._left.size + this._right.size
  }

  get max(): V {
    if (this._left.size === 0) {
      return this._right.max
    }
    if (this._right.size === 0) {
      return this._left.max
    }
    if (this._compareValue(this._left.max, this._right.max) >= 0) {
      return this._left.max
    }
    return this._right.max
  }
}

class _MaxStack<E extends { value: V }, V = E['value']> implements IStack<E> {
  private readonly _stack: E[] = []
  private readonly _maxes: V[] = []
  private readonly _compareValue: CompareFunc<V>

  constructor(compareValue: CompareFunc<V> = (a: any, b: any) => a - b) {
    this._compareValue = compareValue
  }

  append(item: E): void {
    this._stack.push(item)
    if (
      this._maxes.length === 0 ||
      this._compareValue(item.value, this._maxes[this._maxes.length - 1]) >= 0
    ) {
      this._maxes.push(item.value)
    }
  }

  pop(): E | undefined {
    const res = this._stack.pop()
    if (res && res.value === this._maxes[this._maxes.length - 1]) {
      this._maxes.pop()
    }
    return res
  }

  at(index: number): E | undefined {
    return this._stack[index]
  }

  get size(): number {
    return this._stack.length
  }

  get max(): V {
    return this._maxes[this._maxes.length - 1]
  }
}

class MinDeque<E extends { value: V }, V = E['value']> implements IDeque<E> {
  private readonly _left: _MinStack<E, V>
  private readonly _right: _MinStack<E, V>
  private readonly _compareValue: CompareFunc<V>

  constructor(compareValue: CompareFunc<V> = (a: any, b: any) => a - b) {
    this._left = new _MinStack(compareValue)
    this._right = new _MinStack(compareValue)
    this._compareValue = compareValue
  }

  append(item: E): void {
    this._right.append(item)
  }

  appendLeft(item: E): void {
    this._left.append(item)
  }

  pop(): E | undefined {
    if (this._right.size > 0) {
      return this._right.pop()
    }
    const tmp: E[] = []
    const n = this._left.size
    for (let i = 0; i < n; i++) {
      tmp.push(this._left.pop()!)
    }
    const half = n >> 1
    for (let i = half - 1; i >= 0; i--) {
      this._left.append(tmp[i])
    }
    for (let i = half; i < n; i++) {
      this._right.append(tmp[i])
    }
    return this._right.pop()
  }

  popLeft(): E | undefined {
    if (this._left.size > 0) {
      return this._left.pop()
    }
    const tmp: E[] = []
    const n = this._right.size
    for (let i = 0; i < n; i++) {
      tmp.push(this._right.pop()!)
    }
    const half = n >> 1
    for (let i = half - 1; i >= 0; i--) {
      this._right.append(tmp[i])
    }
    for (let i = half; i < n; i++) {
      this._left.append(tmp[i])
    }
    return this._left.pop()
  }

  at(index: number): E | undefined {
    const n = this.size
    if (index < 0) {
      index += n
    }
    if (index < 0 || index >= n) {
      return undefined
    }
    const leftSize = this._left.size
    if (index < leftSize) {
      return this._left.at(leftSize - index - 1)
    }
    return this._right.at(index - leftSize)
  }

  toString(): string {
    const sb: string[] = []
    const n = this.size
    for (let i = 0; i < n; i++) {
      sb.push(JSON.stringify(this.at(i)))
    }
    return `MinDeque(${sb.join(', ')})`
  }

  get size(): number {
    return this._left.size + this._right.size
  }

  get min(): V {
    if (this._left.size === 0) {
      return this._right.min
    }
    if (this._right.size === 0) {
      return this._left.min
    }
    if (this._compareValue(this._left.min, this._right.min) <= 0) {
      return this._left.min
    }
    return this._right.min
  }
}

class _MinStack<E extends { value: V }, V = E['value']> implements IStack<E> {
  private readonly _stack: E[] = []
  private readonly _mins: V[] = []
  private readonly _compareValue: CompareFunc<V>

  constructor(compareValue: CompareFunc<V> = (a: any, b: any) => a - b) {
    this._compareValue = compareValue
  }

  append(item: E): void {
    this._stack.push(item)
    if (
      this._mins.length === 0 ||
      this._compareValue(item.value, this._mins[this._mins.length - 1]) <= 0
    ) {
      this._mins.push(item.value)
    }
  }

  pop(): E | undefined {
    const res = this._stack.pop()
    if (res && res.value === this._mins[this._mins.length - 1]) {
      this._mins.pop()
    }
    return res
  }

  at(index: number): E | undefined {
    return this._stack[index]
  }

  get size(): number {
    return this._stack.length
  }

  get min(): V {
    return this._mins[this._mins.length - 1]
  }
}

if (require.main === module) {
  const maxDeque = new MaxDeque()
  maxDeque.append({ value: 1 })
  maxDeque.append({ value: 3 })
  maxDeque.append({ value: 2 })
  maxDeque.append({ value: 4 })
  maxDeque.append({ value: 5 })
  console.log(maxDeque.toString())
  const res = maxDeque.popLeft()
  console.log(res)
  console.log(maxDeque.toString())
  maxDeque.appendLeft({ value: 6 })
  console.log(maxDeque.toString())
  const res2 = maxDeque.popLeft()
  console.log(res2)
  console.log(maxDeque.toString())
  console.log(maxDeque.max)
  console.log(maxDeque.at(0))
  console.log(maxDeque.size)

  class MyCircularDeque {
    private readonly _queue: IDeque<{ value: number }> = new MinDeque()
    private readonly _k: number
    constructor(k: number) {
      this._k = k
    }

    insertFront(value: number): boolean {
      if (this._queue.size === this._k) {
        return false
      }
      this._queue.appendLeft({ value })
      return true
    }

    insertLast(value: number): boolean {
      if (this._queue.size === this._k) {
        return false
      }
      this._queue.append({ value })
      return true
    }

    deleteFront(): boolean {
      if (this._queue.size === 0) {
        return false
      }
      this._queue.popLeft()
      return true
    }

    deleteLast(): boolean {
      if (this._queue.size === 0) {
        return false
      }
      this._queue.pop()
      return true
    }

    getFront(): number {
      const res = this._queue.at(0)
      return res ? res.value : -1
    }

    getRear(): number {
      const res = this._queue.at(this._queue.size - 1)
      return res ? res.value : -1
    }

    isEmpty(): boolean {
      return this._queue.size === 0
    }

    isFull(): boolean {
      return this._queue.size === this._k
    }
  }

  /**
   * Your MyCircularDeque object will be instantiated and called as such:
   * var obj = new MyCircularDeque(k)
   * var param_1 = obj.insertFront(value)
   * var param_2 = obj.insertLast(value)
   * var param_3 = obj.deleteFront()
   * var param_4 = obj.deleteLast()
   * var param_5 = obj.getFront()
   * var param_6 = obj.getRear()
   * var param_7 = obj.isEmpty()
   * var param_8 = obj.isFull()
   */

  const myMinDeque = new MaxDeque<{ value: string }>((a, b) => a.localeCompare(b))
  myMinDeque.append({ value: 'a' })
  myMinDeque.append({ value: 'c' })
}

export { MaxDeque, MinDeque }

File: 11_动态规划/dp优化/辅助数据结构dp/单调队列优化dp/单调队列Monoqueue/MaxDequeue.py
# 用两个栈模拟一个双端队列，注意到每个栈的前缀最小值是可以 O(1) 修改和询问的
# 当某个栈空的时候将另一个栈分一半给这个已经空的栈（即暴力重构）

# 参考 SlidingWindowAggregationDeque


from typing import Tuple


E = Tuple[int, ...]  # (value, index,...)


class MaxDeque:
    __slots__ = ("_left", "_right")

    def __init__(self):
        self._left = _MaxStack()
        self._right = _MaxStack()

    def append(self, x: "E") -> None:
        self._right.append(x)

    def appendleft(self, x: "E") -> None:
        self._left.append(x)

    def pop(self) -> "E":
        if self._right:
            return self._right.pop()
        tmp = []
        n = len(self._left)
        for _ in range(n):
            tmp.append(self._left.pop())
        half = n // 2
        for i in range(half - 1, -1, -1):
            self._left.append(tmp[i])
        for i in range(half, n):
            self._right.append(tmp[i])
        return self._right.pop()

    def popleft(self) -> "E":
        if self._left:
            return self._left.pop()
        tmp = []
        n = len(self._right)
        for _ in range(n):
            tmp.append(self._right.pop())
        half = n // 2
        for i in range(half - 1, -1, -1):
            self._right.append(tmp[i])
        for i in range(half, n):
            self._left.append(tmp[i])
        return self._left.pop()

    @property
    def max(self) -> int:
        if not self._left:
            return self._right.max
        if not self._right:
            return self._left.max
        return self._left.max if self._left.max > self._right.max else self._right.max

    def __len__(self) -> int:
        return len(self._left) + len(self._right)

    def __getitem__(self, i: int) -> "E":
        n = len(self)
        if i < 0:
            i += n
        if i < 0 or i >= n:
            raise IndexError("deque index out of range")
        if i < len(self._left):
            return self._left[-i - 1]
        return self._right[i - len(self._left)]

    def __repr__(self) -> str:
        sb = []
        for i in range(len(self)):
            sb.append(str(self[i]))
        return f"MaxDeque({', '.join(sb)})"


class _MaxStack:
    __slots__ = ("_stack", "_maxes")

    def __init__(self):
        self._stack = []
        self._maxes = []

    def append(self, x: "E") -> None:
        self._stack.append(x)
        if not self._maxes or x[0] >= self._maxes[-1]:
            self._maxes.append(x[0])

    def pop(self) -> "E":
        res = self._stack.pop()
        if res[0] == self._maxes[-1]:
            self._maxes.pop()
        return res

    def top(self) -> "E":
        return self._stack[-1]

    @property
    def max(self) -> int:
        return self._maxes[-1]

    def __len__(self) -> int:
        return len(self._stack)

    def __getitem__(self, i: int) -> "E":
        return self._stack[i]

    def __repr__(self) -> str:
        return f"{self._stack}"


E = Tuple[int, ...]  # (value, index,...)


class MinDeque:
    __slots__ = ("_left", "_right")

    def __init__(self):
        self._left = _MinStack()
        self._right = _MinStack()

    def append(self, x: "E") -> None:
        self._right.append(x)

    def appendleft(self, x: "E") -> None:
        self._left.append(x)

    def pop(self) -> "E":
        if self._right:
            return self._right.pop()
        tmp = []
        n = len(self._left)
        for _ in range(n):
            tmp.append(self._left.pop())
        half = n // 2
        for i in range(half - 1, -1, -1):
            self._left.append(tmp[i])
        for i in range(half, n):
            self._right.append(tmp[i])
        return self._right.pop()

    def popleft(self) -> "E":
        if self._left:
            return self._left.pop()
        tmp = []
        n = len(self._right)
        for _ in range(n):
            tmp.append(self._right.pop())
        half = n // 2
        for i in range(half - 1, -1, -1):
            self._right.append(tmp[i])
        for i in range(half, n):
            self._left.append(tmp[i])
        return self._left.pop()

    @property
    def min(self) -> int:
        if not self._left:
            return self._right.min
        if not self._right:
            return self._left.min
        return self._left.min if self._left.min < self._right.min else self._right.min

    def __len__(self) -> int:
        return len(self._left) + len(self._right)

    def __getitem__(self, i: int) -> "E":
        n = len(self)
        if i < 0:
            i += n
        if i < 0 or i >= n:
            raise IndexError("deque index out of range")
        if i < len(self._left):
            return self._left[-i - 1]
        return self._right[i - len(self._left)]

    def __repr__(self) -> str:
        sb = []
        for i in range(len(self)):
            sb.append(str(self[i]))
        return f"MinDeque({', '.join(sb)})"


class _MinStack:
    __slots__ = ("_stack", "_mins")

    def __init__(self):
        self._stack = []
        self._mins = []

    def append(self, x: "E") -> None:
        self._stack.append(x)
        if not self._mins or x[0] <= self._mins[-1]:
            self._mins.append(x[0])

    def pop(self) -> "E":
        res = self._stack.pop()
        if res[0] == self._mins[-1]:
            self._mins.pop()
        return res

    def top(self) -> "E":
        return self._stack[-1]

    @property
    def min(self) -> int:
        return self._mins[-1]

    def __len__(self) -> int:
        return len(self._stack)

    def __getitem__(self, i: int) -> "E":
        return self._stack[i]

    def __repr__(self) -> str:
        return f"{self._stack}"


if __name__ == "__main__":
    maxDeque = MaxDeque()
    maxDeque.append((1, 0))
    maxDeque.append((3, 1))
    maxDeque.append((2, 2))
    maxDeque.append((4, 3))
    maxDeque.append((5, 4))
    print(maxDeque)
    res = maxDeque.popleft()
    assert res == (1, 0)
    print(res)
    print(maxDeque)
    maxDeque.appendleft((6, 5))
    print(maxDeque)
    res = maxDeque.popleft()
    assert res == (6, 5)
    print(res)
    print(maxDeque)
    assert maxDeque.max == 5
    assert maxDeque[0] == (3, 1)
    assert len(maxDeque) == 4

    minDeque = MinDeque()
    minDeque.append((1, 0))
    minDeque.append((3, 1))
    minDeque.append((2, 2))
    minDeque.append((4, 3))
    minDeque.append((5, 4))
    print(minDeque)
    res = minDeque.popleft()
    assert res == (1, 0)
    print(res)
    print(minDeque)
    minDeque.appendleft((6, 5))
    print(minDeque)
    res = minDeque.popleft()
    assert res == (6, 5)
    print(res)
    print(minDeque)
    assert minDeque.min == 2
    assert minDeque[0] == (3, 1)
    assert len(minDeque) == 4
